### PR TITLE
Simplify CommandBox #505

### LINKED ISSUE #505: Simplify CommandBox
CommandBox is too ambitious. It is executing on `MainWindow`'s `model` and updating the `ResultDisplay` by raising a `NewResultAvailableEvent`.

These ought to be the responsibility of the `MainWindow`; MainWindow should operate on its own `model`. Also, being the parent container of both `CommandBox` and `ResultDisplay`, it should be the one coordinating the update of `ResultDisplay`

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Fixes #505

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/4] [Add CommandBoxDelegate and teach MainWindow to implement it](https://github.com/se-edu/addressbook-level4/pull/506/commits/81fc55cdd9533d4af185925d3d4eb21b1da58a66)
* [v1 2/4] [CommandBox: Remove \#logic](https://github.com/se-edu/addressbook-level4/pull/506/commits/5586e3df21ccc06e98b80b32e09c5ee031210d52)
* [v1 3/4] [CommandBox: Remove raising of NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/baaef307b97d86324d9e7d8793802e23bea58dc0)
* [v1 4/4] [Remove NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/4b25b297c122cb596efef153acdf5cce6afc9229)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...4b25b297c122cb596efef153acdf5cce6afc9229))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/506/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: src/main/java/seedu/address/ui/CommandBoxDelegate.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package seedu.address.ui;
+
+import seedu.address.logic.commands.CommandResult;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.logic.parser.exceptions.ParseException;
+
+/**
+ * Executes the user input entered into the {@code CommandBox} on its behalf.
+ */
+public interface CommandBoxDelegate {
```
Comment (yamgent) [inline src/main/java/seedu/address/ui/CommandBoxDelegate.java]: `ExecuteDelegate`?

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/4] [Add CommandBoxDelegate and teach MainWindow to implement it](https://github.com/se-edu/addressbook-level4/pull/506/commits/0c92ab91a735608db874487cb7cec251fdbc2329)
* [v2 2/4] [CommandBox: Remove CommandBox\#logic](https://github.com/se-edu/addressbook-level4/pull/506/commits/fdb4c3728f4ae5e3b16e57d7ec7e044a59f36723)
* [v2 3/4] [CommandBox: Remove raising of NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/aa1842a0586844a2c0aca42fae8d3694d3e4836d)
* [v2 4/4] [Remove NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/f29b7d41dc619262fbb95c98adcda2f4077a53e1)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...f29b7d41dc619262fbb95c98adcda2f4077a53e1)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/6ba459c2dacb1c50ba36e9ff289e74cb280acb26~1...6ba459c2dacb1c50ba36e9ff289e74cb280acb26))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/506/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
Ignore v2, I forgot to change a commit message, sorry!

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/4] [Add ExecuteDelegate and teach MainWindow to implement it](https://github.com/se-edu/addressbook-level4/pull/506/commits/7542f45a0b3fe65089a2d84c4db5b37501498ecf)
* [v3 2/4] [CommandBox: Remove CommandBox\#logic](https://github.com/se-edu/addressbook-level4/pull/506/commits/e7323c2e7a6a420fdaed8a38fb29d615bc222f58)
* [v3 3/4] [CommandBox: Remove raising of NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/45168c822e7c5496288bedcac71303ec6b80dc06)
* [v3 4/4] [Remove NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/e86b201ae3906b8e6d9939798996ba73c1f898c5)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...e86b201ae3906b8e6d9939798996ba73c1f898c5)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/18357f73528535926e266be9ecad18dbd1e3e85b~1...18357f73528535926e266be9ecad18dbd1e3e85b))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/506/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by Zhiyuan-Amos** [comment]
@yamgent @limmlingg for your review; need to merge quickly this cos there's other dependencies :>

File: src/main/java/seedu/address/ui/CommandBox.java [hunk 2]
```diff
@@ -21,14 +21,16 @@
     private static final String FXML = "CommandBox.fxml";
 
     private final Logger logger = LogsCenter.getLogger(CommandBox.class);
+    private ExecuteDelegate delegate;
```
Comment (yamgent) [inline src/main/java/seedu/address/ui/CommandBox.java]: Variable name should be `executeDelegate`?

File: src/main/java/seedu/address/ui/ExecuteDelegate.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package seedu.address.ui;
+
+import seedu.address.logic.commands.CommandResult;
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.logic.parser.exceptions.ParseException;
+
+/**
+ * Executes the user input entered into the {@code CommandBox} on its behalf.
```
Comment (yamgent) [inline src/main/java/seedu/address/ui/ExecuteDelegate.java]: Not sure whether we should mention about how it is being used.

**Comment by CanIHasReview** [comment]
# v4

@Zhiyuan-Amos submitted v4 for review.

* [v4 1/4] [Add ExecuteDelegate and teach MainWindow to implement it](https://github.com/se-edu/addressbook-level4/pull/506/commits/4efb44d17220a331060939fc1017ee3ed4bc7bd3)
* [v4 2/4] [CommandBox: Remove CommandBox\#logic](https://github.com/se-edu/addressbook-level4/pull/506/commits/894470080ed1ac2416a7c76f8e3fa509404a03fe)
* [v4 3/4] [CommandBox: Remove raising of NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/3fad77011d4dafaa134274911e39fabc9ecb4002)
* [v4 4/4] [Remove NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/dd3bad56ddeb92195f2660b6b61b83bb0c6b57f9)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...dd3bad56ddeb92195f2660b6b61b83bb0c6b57f9)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/3d5aaf28453307db3ab0c65749721f9a88eb350d~1...3d5aaf28453307db3ab0c65749721f9a88eb350d))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/506/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by CanIHasReview** [comment]
# v5

@Zhiyuan-Amos submitted v5 for review.

* [v5 1/5] [Create ExecuteDelegate](https://github.com/se-edu/addressbook-level4/pull/506/commits/e815f3b63ee66742dfacb160bf994c47952f7bc6)
* [v5 2/5] [Teach MainWindow to implement ExecuteDelegate](https://github.com/se-edu/addressbook-level4/pull/506/commits/c3bc250f8343431660ee8aa6af1d2938f7968557)
* [v5 3/5] [CommandBox: Remove CommandBox\#logic](https://github.com/se-edu/addressbook-level4/pull/506/commits/0794f43bcb9406851b73aef53aa1652fb10f3d1e)
* [v5 4/5] [CommandBox: Remove raising of NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/5d6477f5b1b43fa387d99dcbf167ca06148a580d)
* [v5 5/5] [Remove NewResultAvailableEvent](https://github.com/se-edu/addressbook-level4/pull/506/commits/7d18849abcdb1a8bc5dcbe5306c5e2a241b07617)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/5cc85fcf11b40d1660d0b6be20a3ef90bc71c1c1...7d18849abcdb1a8bc5dcbe5306c5e2a241b07617)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/635d110e29f82911230906ca1a946efd9165626e~1...635d110e29f82911230906ca1a946efd9165626e))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/506/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by damithc** [comment]
As discussed, this PR is a step towards a different GUI architecture where GUI widgets are mostly passive and a 'supervising controller' (see [Martin Fowler's article on GUI architectures](https://martinfowler.com/eaaDev/uiArchs.html)) handles most of the GUI logic. The current architecture on the other hand has a different philosophy where each widget handles it's own logic in response to user actions on it or events that it is interested in. i.e. there is no 'supervising controller'. I'm sure each approach has pros and cons and it is also possible that the 'supervising controller' approach is superior. However, I don't think we should start moving towards a new GUI architecture unless we are sure we want to go all the way. Shall we put this on hold until then?

**Comment by chao1995** [comment]
@Zhiyuan-Amos Please request review from me again once we decide this new architecture is what we want to use.

**Comment by Zhiyuan-Amos** [comment]
@chao1995 noted with thanks :>

**Comment by Zhiyuan-Amos** [comment]
Closing this PR: Once we begin work on this PR we will have to revamp the entire architecture which requires quite a bit of work. I will leave this to future CS3282 students / interns to finish.

**Pull request closed without merging**